RuntimeException with 2 or more presenters

If there are two or more presenters and the view is destroyed, when the getView () method is invoked the proxy works for a presenter but not for more.
If two or more are implemented, only the last one works.

Example:
```
_public class CharacterDetailsActivity extends MarvelActivity
    implements **CharacterCrashDetailsPresenter.View**, CharacterDetailsPresenter.View {...}_

Process: com.karumi.rosie.sample, PID: 20529
                                                   java.lang.RuntimeException: Internal error invoking the success object
                                                       at com.karumi.rosie.domain.usecase.RosieUseCase$1.run(RosieUseCase.java:114)
                                                       at android.os.Handler.handleCallback(Handler.java:739)
                                                       at android.os.Handler.dispatchMessage(Handler.java:95)
                                                       at android.os.Looper.loop(Looper.java:234)
                                                       at android.app.ActivityThread.main(ActivityThread.java:5526)
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                    Caused by: java.lang.reflect.InvocationTargetException
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.karumi.rosie.domain.usecase.RosieUseCase$1.run(RosieUseCase.java:112)
                                                       at android.os.Handler.handleCallback(Handler.java:739)
                                                       at android.os.Handler.dispatchMessage(Handler.java:95)
                                                       at android.os.Looper.loop(Looper.java:234)
                                                       at android.app.ActivityThread.main(ActivityThread.java:5526)
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                    Caused by: java.lang.ClassCastException: $Proxy5 cannot be cast to com.karumi.rosie.sample.characters.view.presenter.CharacterCrashDetailsPresenter$View
                                                       at com.karumi.rosie.sample.characters.view.presenter.CharacterCrashDetailsPresenter$2.onCharacterDetailsLoaded(CharacterCrashDetailsPresenter.java:62)
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.karumi.rosie.domain.usecase.RosieUseCase$1.run(RosieUseCase.java:112)
                                                       at android.os.Handler.handleCallback(Handler.java:739)
                                                       at android.os.Handler.dispatchMessage(Handler.java:95)
                                                       at android.os.Looper.loop(Looper.java:234)
                                                       at android.app.ActivityThread.main(ActivityThread.java:5526)
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
```
Thanks in advance.

--- a/.travis.yml
+++ b/.travis.yml
@@ -9,16 +9,6 @@ android:
     - extra-android-support
     - extra-google-m2repository
     - extra-android-m2repository
-    - sys-img-armeabi-v7a-android-21
-
-before_script:
-  - echo no | android create avd --force -n test -t android-21 --abi armeabi-v7a
-  - emulator -avd test -no-skin -no-audio -no-window &
-  - android-wait-for-emulator
-  - adb shell settings put global window_animation_scale 0 &
-  - adb shell settings put global transition_animation_scale 0 &
-  - adb shell settings put global animator_duration_scale 0 &
-  - adb shell input keyevent 82 &
 
 script:
-  - ./gradlew checkstyle build connectedCheck -Dpre-dex=false -Pcom.android.build.threadPoolSize=1 -Dorg.gradle.parallel=false -Dorg.gradle.jvmargs="-Xms512m -Xmx512m" -Dorg.gradle.daemon=false
+  - ./gradlew checkstyle build -Dpre-dex=false -Pcom.android.build.threadPoolSize=1 -Dorg.gradle.parallel=false -Dorg.gradle.jvmargs="-Xms512m -Xmx512m" -Dorg.gradle.daemon=false

--- a/rosie/src/main/java/com/karumi/rosie/view/RosiePresenter.java
+++ b/rosie/src/main/java/com/karumi/rosie/view/RosiePresenter.java
@@ -33,6 +33,11 @@ import java.util.List;
  */
 public class RosiePresenter<T extends RosiePresenter.View> {
 
+  private static final InvocationHandler emptyHandler = new InvocationHandler() {
+    @Override public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+      return null;
+    }
+  };
   private final UseCaseHandler useCaseHandler;
   private final List<UseCaseCall> useCaseCalls = new LinkedList<>();
   private final List<OnErrorCallback> globalOnErrorCallbacks = new LinkedList<>();
@@ -87,7 +92,6 @@ public class RosiePresenter<T extends RosiePresenter.View> {
     return useCaseCall;
   }
 
-
   /**
    * Returns the view configured in the presenter which real implementation is an Activity or
    * Fragment using this presenter.
@@ -124,29 +128,34 @@ public class RosiePresenter<T extends RosiePresenter.View> {
    * Changes the current view instance with a dynamic proxy to avoid real UI updates.
    */
   void resetView() {
-    final Class<?> viewClass = getViewInterfaceClass();
-    InvocationHandler emptyHandler = new InvocationHandler() {
-      @Override
-      public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-        return null;
-      }
-    };
-    ClassLoader classLoader = viewClass.getClassLoader();
-    Class[] interfaces = new Class[1];
-    interfaces[0] = viewClass;
-    this.view = (T) Proxy.newProxyInstance(classLoader, interfaces, emptyHandler);
+    final Class[] viewClasses = getViewInterfaceClasses();
+    ClassLoader classLoader = viewClasses[0].getClassLoader();
+    this.view = (T) Proxy.newProxyInstance(classLoader, viewClasses, emptyHandler);
   }
 
-  private Class<?> getViewInterfaceClass() {
-    Class<?> interfaceClass = null;
+  private Class<?>[] getViewInterfaceClasses() {
+    List<Class<?>> interfaceClasses = new LinkedList<>();
     Class<?>[] interfaces = this.view.getClass().getInterfaces();
     for (int i = 0; i < interfaces.length; i++) {
       Class<?> interfaceCandidate = interfaces[i];
-      if (RosiePresenter.View.class.isAssignableFrom(interfaceCandidate)) {
-        interfaceClass = interfaceCandidate;
+      if (RosiePresenter.View.class.isAssignableFrom(
+          interfaceCandidate)) {
+        interfaceClasses.add(interfaceCandidate);
       }
     }
-    return interfaceClass;
+    return interfaceClasses.toArray(new Class[interfaceClasses.size()]);
+  }
+
+  private boolean canDowncastInstance(Class<?> candidate) {
+    ClassLoader classLoader = candidate.getClassLoader();
+    Class[] interfaces = new Class[1];
+    interfaces[0] = candidate;
+    try {
+      T safeCast = (T) Proxy.newProxyInstance(classLoader, interfaces, emptyHandler);
+    } catch (Exception e) {
+      return false;
+    }
+    return true;
   }
 
   private void registerGlobalErrorCallback() {

--- a/rosie/src/main/java/com/karumi/rosie/view/RosiePresenter.java
+++ b/rosie/src/main/java/com/karumi/rosie/view/RosiePresenter.java
@@ -20,9 +20,6 @@ import com.karumi.rosie.domain.usecase.RosieUseCase;
 import com.karumi.rosie.domain.usecase.UseCaseCall;
 import com.karumi.rosie.domain.usecase.UseCaseHandler;
 import com.karumi.rosie.domain.usecase.error.OnErrorCallback;
-import java.lang.reflect.InvocationHandler;
-import java.lang.reflect.Method;
-import java.lang.reflect.Proxy;
 import java.util.LinkedList;
 import java.util.List;
 
@@ -33,11 +30,7 @@ import java.util.List;
  */
 public class RosiePresenter<T extends RosiePresenter.View> {
 
-  private static final InvocationHandler emptyHandler = new InvocationHandler() {
-    @Override public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-      return null;
-    }
-  };
+  private final RosieViewProxyGenerator<T> rosieViewProxyGenerator = new RosieViewProxyGenerator<>();
   private final UseCaseHandler useCaseHandler;
   private final List<UseCaseCall> useCaseCalls = new LinkedList<>();
   private final List<OnErrorCallback> globalOnErrorCallbacks = new LinkedList<>();
@@ -128,34 +121,7 @@ public class RosiePresenter<T extends RosiePresenter.View> {
    * Changes the current view instance with a dynamic proxy to avoid real UI updates.
    */
   void resetView() {
-    final Class[] viewClasses = getViewInterfaceClasses();
-    ClassLoader classLoader = viewClasses[0].getClassLoader();
-    this.view = (T) Proxy.newProxyInstance(classLoader, viewClasses, emptyHandler);
-  }
-
-  private Class<?>[] getViewInterfaceClasses() {
-    List<Class<?>> interfaceClasses = new LinkedList<>();
-    Class<?>[] interfaces = this.view.getClass().getInterfaces();
-    for (int i = 0; i < interfaces.length; i++) {
-      Class<?> interfaceCandidate = interfaces[i];
-      if (RosiePresenter.View.class.isAssignableFrom(
-          interfaceCandidate)) {
-        interfaceClasses.add(interfaceCandidate);
-      }
-    }
-    return interfaceClasses.toArray(new Class[interfaceClasses.size()]);
-  }
-
-  private boolean canDowncastInstance(Class<?> candidate) {
-    ClassLoader classLoader = candidate.getClassLoader();
-    Class[] interfaces = new Class[1];
-    interfaces[0] = candidate;
-    try {
-      T safeCast = (T) Proxy.newProxyInstance(classLoader, interfaces, emptyHandler);
-    } catch (Exception e) {
-      return false;
-    }
-    return true;
+    this.view = rosieViewProxyGenerator.generate(view);
   }
 
   private void registerGlobalErrorCallback() {

--- a/rosie/src/main/java/com/karumi/rosie/view/RosieViewProxyGenerator.java
+++ b/rosie/src/main/java/com/karumi/rosie/view/RosieViewProxyGenerator.java
@@ -1,0 +1,50 @@
+/*
+ * Copyright (C) 2015 Karumi.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.karumi.rosie.view;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.LinkedList;
+import java.util.List;
+
+class RosieViewProxyGenerator<T> {
+
+  private static final InvocationHandler emptyHandler = new InvocationHandler() {
+    @Override public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+      return null;
+    }
+  };
+
+  T generate(Object view) {
+    final Class[] viewClasses = getViewInterfaceClasses(view);
+    ClassLoader classLoader = viewClasses[0].getClassLoader();
+    return (T) Proxy.newProxyInstance(classLoader, viewClasses, emptyHandler);
+  }
+
+  private Class<?>[] getViewInterfaceClasses(Object view) {
+    List<Class<?>> interfaceClasses = new LinkedList<>();
+    Class<?>[] interfaces = view.getClass().getInterfaces();
+    for (int i = 0; i < interfaces.length; i++) {
+      Class<?> interfaceCandidate = interfaces[i];
+      if (RosiePresenter.View.class.isAssignableFrom(interfaceCandidate)) {
+        interfaceClasses.add(interfaceCandidate);
+      }
+    }
+    return interfaceClasses.toArray(new Class[interfaceClasses.size()]);
+  }
+}

--- a/rosie/src/main/java/com/karumi/rosie/view/RosieViewProxyGenerator.java
+++ b/rosie/src/main/java/com/karumi/rosie/view/RosieViewProxyGenerator.java
@@ -24,16 +24,21 @@ import java.util.List;
 
 class RosieViewProxyGenerator<T> {
 
-  private static final InvocationHandler emptyHandler = new InvocationHandler() {
+  private static final InvocationHandler EMPTY_HANDLER = new InvocationHandler() {
     @Override public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
       return null;
     }
   };
 
+  /**
+   * Generates a proxy instance for classes implementing RosiePresenter.View. The instance
+   * generated uses a default handler so it will generate any side effect when invoking this
+   * instance but it will be equivalent from the type point of view.
+   */
   T generate(Object view) {
     final Class[] viewClasses = getViewInterfaceClasses(view);
     ClassLoader classLoader = viewClasses[0].getClassLoader();
-    return (T) Proxy.newProxyInstance(classLoader, viewClasses, emptyHandler);
+    return (T) Proxy.newProxyInstance(classLoader, viewClasses, EMPTY_HANDLER);
   }
 
   private Class<?>[] getViewInterfaceClasses(Object view) {

--- a/rosie/src/test/java/com/karumi/rosie/view/RosieViewProxyGeneratorTest.java
+++ b/rosie/src/test/java/com/karumi/rosie/view/RosieViewProxyGeneratorTest.java
@@ -1,0 +1,61 @@
+/*
+ * Copyright (C) 2015 Karumi.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.karumi.rosie.view;
+
+import org.junit.Test;
+
+import static junit.framework.Assert.assertFalse;
+import static junit.framework.Assert.assertTrue;
+
+public class RosieViewProxyGeneratorTest {
+
+  private final RosieViewProxyGenerator<C> generator = new RosieViewProxyGenerator<>();
+  private final C anyClassExtendingInterfaces = new C();
+
+  @Test
+  public void shouldGenerateAClassImplementingBothInterfacesExtendingFromRosieViewUsingAnEmptyHandlerWeCanInvokeWithoutCrashing() {
+    A a = generator.generate(anyClassExtendingInterfaces);
+    B b = generator.generate(anyClassExtendingInterfaces);
+
+    assertTrue(a instanceof A);
+    assertTrue(a instanceof B);
+    assertTrue(b instanceof A);
+    assertTrue(b instanceof B);
+  }
+
+  @Test public void shouldNotGenerateTheClassUsingTheInstancesNotExtendingFromRosieView() {
+    A a = generator.generate(anyClassExtendingInterfaces);
+
+    assertFalse(a instanceof NotARosieView);
+  }
+
+  interface A extends RosiePresenter.View {
+
+  }
+
+  interface B extends RosiePresenter.View {
+
+  }
+
+  interface NotARosieView {
+
+  }
+
+  class C implements A, B, NotARosieView {
+
+  }
+}

--- a/rosie/src/test/java/com/karumi/rosie/view/RosieViewProxyGeneratorTest.java
+++ b/rosie/src/test/java/com/karumi/rosie/view/RosieViewProxyGeneratorTest.java
@@ -27,7 +27,7 @@ public class RosieViewProxyGeneratorTest {
   private final C anyClassExtendingInterfaces = new C();
 
   @Test
-  public void shouldGenerateAClassImplementingBothInterfacesExtendingFromRosieViewUsingAnEmptyHandlerWeCanInvokeWithoutCrashing() {
+  public void shouldGenerateAClassImplementingBothInterfacesExtendingFromRosieView() {
     A a = generator.generate(anyClassExtendingInterfaces);
     B b = generator.generate(anyClassExtendingInterfaces);
 

